src/content/actions/receive-profile.js

```diff
@@ -167,46 +167,83 @@ export function assignTaskTracerNames(addressIndices: number[], symbolNames: str
+    onMergeFunctions: (threadIndex: ThreadIndex, oldFuncToNewFuncMap: FuncToFuncMap) => {
+      dispatch(mergeFunctions(threadIndex, oldFuncToNewFuncMap));
+    },
+    onGotFuncNames: (threadIndex: ThreadIndex, funcIndices: IndexIntoFuncTable[], funcNames: string[]) => {
+      dispatch(assignFunctionNames(threadIndex, funcIndices, funcNames));
+    },
+    onGotTaskTracerNames: (addressIndices, symbolNames) => {
+      dispatch(assignTaskTracerNames(addressIndices, symbolNames));
+    },
+  });
+  dispatch(doneSymbolicating());
+}
+
+export function errorReceivingProfileFromAddon(error: Error) {
+  return {
+    type: 'ERROR_RECEIVING_PROFILE_FROM_ADDON',
+    error,
+  };
+}
+
 export function retrieveProfileFromAddon(): ThunkAction {
-  return dispatch => {
-    dispatch(waitingForProfileFromAddon());
-
-    // XXX use Promise.race with a 5 second timeout promise to show an error message
-    window.geckoProfilerPromise.then(geckoProfiler => {
-      // XXX update state to show that we're connected to the profiler addon
-      geckoProfiler.getProfile().then(rawGeckoProfile => {
-        const profile = processProfile(rawGeckoProfile);
-
-        dispatch(receiveProfileFromAddon(profile));
-
-        const symbolStore = new SymbolStore('perf-html-async-storage', {
-          requestSymbolTable: (debugName, breakpadId) => {
-            const requestedLib = { debugName, breakpadId };
-            dispatch(requestingSymbolTable(requestedLib));
-            return geckoProfiler.getSymbolTable(debugName, breakpadId).then(symbolTable => {
-              dispatch(receivedSymbolTableReply(requestedLib));
-              return symbolTable;
-            }, error => {
-              dispatch(receivedSymbolTableReply(requestedLib));
-              throw error;
-            });
-          },
-        });
-
-        dispatch(startSymbolicating());
-        symbolicateProfile(profile, symbolStore, {
-          onMergeFunctions: (threadIndex: ThreadIndex, oldFuncToNewFuncMap: FuncToFuncMap) => {
-            dispatch(mergeFunctions(threadIndex, oldFuncToNewFuncMap));
-          },
-          onGotFuncNames: (threadIndex: ThreadIndex, funcIndices: IndexIntoFuncTable[], funcNames: string[]) => {
-            dispatch(assignFunctionNames(threadIndex, funcIndices, funcNames));
-          },
-          onGotTaskTracerNames: (addressIndices, symbolNames) => {
-            dispatch(assignTaskTracerNames(addressIndices, symbolNames));
-          },
-        }).then(() => dispatch(doneSymbolicating()));
-      });
-    });
+  return async dispatch => {
+    try {
+      const geckoProfiler = await getGeckoProfiler();
+      const [profile, symbolStore] = await Promise.all([
+        getProfileFromAddon(dispatch, geckoProfiler),
+        getSymbolStore(dispatch, geckoProfiler),
+      ]);
+      await doSymbolicateProfile(dispatch, profile, symbolStore);
+    } catch (error) {
+      dispatch(errorReceivingProfileFromAddon(error));
+      throw error;
+    }
```

Comment:
I decided to break the function into several subfunctions to make it more readable and manageable.